arch/arm/src/armv7-a/cp15_clean_dcache_all.S

```diff
@@ -86,36 +86,34 @@
  *
  ****************************************************************************/
 
-	.globl	cp15_flush_dcache_all
-	.type	cp15_flush_dcache_all, function
+	.globl	cp15_clean_dcache_all
+	.type	cp15_clean_dcache_all, function
 
-cp15_flush_dcache_all:
+cp15_clean_dcache_all:
+
+	mrc		CP15_CCSIDR(r1)			/* Read the Cache Size Identification Register */
 
-	mrc		CP15_CCSIDR(r0)			/* Read the Cache Size Identification Register */
 	ldr		r3, =0x7fff			/* Isolate the NumSets field (bits 13-27) */
-	and		r0, r3, r0, lsr #13		/* r0=NumSets (number of sets - 1) */
+	and		r0, r3, r1, lsr #13		/* r0=NumSets (number of sets - 1) */
 
 	ldr		r3, =0x3ff			/* Isolate the way field (bits 3-12) */
-	add		r4, r3, r0, lsr #3		/* r4=(number of ways - 1) */
+	and		r1, r3, r1, lsr #3		/* r1=(number of ways - 1) */
 
-	mov		r1, #0				/* r1 = way loop counter */
 way_loop:
-
-	mov		r3, #0				/* r3 = set loop counter */
+	mov		r3, r0				/* Init Sets */
 set_loop:
 	mov		r2, r1, lsl #30			/* r2 = way loop counter << 30 */
 	orr		r2, r3, lsl #5			/* r2 = set/way cache operation format */
-	mcr		CP15_DCCISW(r2)			/* Data Cache Invalidate by Set/Way */
-	add		r3, r3, #1			/* Increment set counter */
-	cmp		r0, r3				/* Last set? */
+	mcr		CP15_DCCSW(r2)			/* Data Cache Clean by Set/Way */
+	sub		r3, r3, #1			/* Subtraction set counter */
+	cmp		r3, #0				/* Last set? */
 	bne		set_loop			/* Keep looping if not */
 
-	add		r1, r1, #1			/* Increment the way counter */
-	cmp		r4, r1				/* Last way */
+	sub		r1, r1, #1			/* Subtraction the way counter */
+	cmp		r1, #0				/* Last way? */
```

Comment:
@zhuyanlinzyl 
Is this condition correct?
If the number of ways equals 4 (i.e. r1 = 3), it hits only 3times.

Reply:
@masayuki2009  

Yes, `sub` and `bne` only loop 3times.
I use `subs` and `bcs` , It should loop 4times.

I update the patch. please review agian.